gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/MatcherMaximumVersion.java

```diff
@@ -0,0 +1,39 @@
+/*
+ * (c) Copyright 2018 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.gradle.dist;
+
+import com.palantir.sls.versions.OrderableSlsVersion;
+import com.palantir.sls.versions.SlsVersionMatcher;
+import java.util.function.Function;
+
+/**
+ * {@link MaximumVersion} implemented in terms of a {@link SlsVersionMatcher}.
+ */
+final class MatcherMaximumVersion extends MaximumVersion {
```

Comment:
Possibly worth making these two variants inner classes of MaximumVersion?  I think the javadoc should actually just give an example rather than repeating classnames, e.g.:

> Matchers may contain wildcards, e.g. 1.X.X